File: ios/DesMoinesInsider/Views/Components/PremiumGate.swift
import SwiftUI

/// A reusable overlay that locks premium content behind a subscription tier.
/// Shows a blurred preview of the content with an upgrade prompt.
///
/// Usage:
/// ```swift
/// PremiumGate(requiredTier: .insider, feature: "Advanced Filters") {
///     AdvancedFilterContent()
/// }
/// ```
struct PremiumGate<Content: View>: View {
    let requiredTier: SubscriptionTier
    let feature: String
    /// Optional tailored paywall context. When nil, a generic context is built
    /// from `requiredTier` + `feature` (IOS-SUB-010).
    var context: PaywallContext? = nil
    @ViewBuilder let content: () -> Content

    @State private var storeKit = StoreKitService.shared
    @State private var showSubscription = false

    private var hasAccess: Bool {
        switch requiredTier {
        case .free:
            return true
        case .insider:
            return storeKit.currentTier == .insider || storeKit.currentTier == .vip
        case .vip:
            return storeKit.currentTier == .vip
        }
    }

    var body: some View {
        if hasAccess {
            content()
        } else {
            lockedOverlay
        }
    }

    private var lockedOverlay: some View {
        VStack(spacing: 12) {
            HStack(spacing: 8) {
                Image(systemName: "lock.fill")
                    .font(.subheadline)
                    .foregroundStyle(.orange)

                Text("\(requiredTier.displayName) Feature")
                    .font(.subheadline.weight(.semibold))
                    .accessibilityAddTraits(.isHeader)

                Spacer()
            }

            Text(feature)
                .font(.caption)
                .foregroundStyle(.secondary)
                .frame(maxWidth: .infinity, alignment: .leading)

            Button {
                showSubscription = true
            } label: {
                Text("Unlock with \(requiredTier.displayName)")
                    .font(.caption.weight(.semibold))
                    // ≥44pt tap target (IOS-AUDIT-UX-012) without changing the type size.
                    .frame(maxWidth: .infinity, minHeight: 44)
                    .background(Color.accentColor, in: RoundedRectangle(cornerRadius: 10))
                    .foregroundStyle(.white)
            }
            .accessibilityLabel("Upgrade to \(requiredTier.displayName) to unlock \(feature)")
        }
        .padding(14)
        .background(Color(.systemGray6), in: RoundedRectangle(cornerRadius: 14))
        .overlay(
            RoundedRectangle(cornerRadius: 14)
                .stroke(Color.orange.opacity(0.2), lineWidth: 1)
        )
        .sheet(isPresented: $showSubscription) {
            PaywallView(context: context ?? .generic(tier: requiredTier, feature: feature))
        }
        // IOS-AUDIT-UX-052 AC4. Without this VoiceOver walks four separate
        // elements - a lock glyph, a tier heading, the feature text, then the
        // button - and the first three are context for the fourth. `.contain`
        // rather than `.combine` keeps the button independently focusable and
        // tappable, which `.combine` would flatten away.
        .accessibilityElement(children: .contain)
        .accessibilityLabel("\(requiredTier.displayName) feature: \(feature)")
    }
}

/// Inline premium lock label used for individual rows/buttons.
/// Shows a lock icon and "Insider" badge next to the label.
struct PremiumBadge: View {
    var tier: SubscriptionTier = .insider

    var body: some View {
        HStack(spacing: 4) {
            Image(systemName: "lock.fill")
                .font(.system(size: 9))
                .accessibilityHidden(true)
            Text(tier.displayName)
                .font(.system(size: 10, weight: .semibold))
        }
        .foregroundStyle(.white)
        .padding(.horizontal, 7)
        .padding(.vertical, 3)
        .background(tier == .vip ? Color.purple : Color.orange, in: Capsule())
        // Read as one element: "Insider feature, locked" (IOS-AUDIT-UX-012).
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(tier.displayName) feature, locked")
    }
}

#Preview("Locked Gate") {
    PremiumGate(requiredTier: .insider, feature: "Advanced search filters including distance, price range, and rating") {
        Text("Premium Content Here")
    }
    .padding()
}

#Preview("Badge") {
    VStack(spacing: 12) {
        PremiumBadge(tier: .insider)
        PremiumBadge(tier: .vip)
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/ui/components/PremiumGate.kt
package com.desmoines.aipulse.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.desmoines.aipulse.data.model.SubscriptionTier
import com.desmoines.aipulse.ui.theme.BrandOrange
import com.desmoines.aipulse.ui.theme.DesMoinesInsiderTheme

/**
 * A reusable overlay that locks premium content behind a subscription tier.
 * Shows a locked state with feature name, tier badge, and unlock button.
 * Mirrors iOS PremiumGate.swift.
 *
 * @param requiredTier The minimum tier required to access the content
 * @param feature Description of the locked feature
 * @param currentTier The user's current subscription tier
 * @param onUpgradeClick Callback when "Unlock" button is tapped
 * @param content The premium content to show when user has access
 */
@Composable
fun PremiumGate(
    requiredTier: SubscriptionTier,
    feature: String,
    currentTier: SubscriptionTier,
    onUpgradeClick: () -> Unit,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit
) {
    val hasAccess = when (requiredTier) {
        SubscriptionTier.FREE -> true
        SubscriptionTier.INSIDER -> currentTier == SubscriptionTier.INSIDER || currentTier == SubscriptionTier.VIP
        SubscriptionTier.VIP -> currentTier == SubscriptionTier.VIP
    }

    if (hasAccess) {
        content()
    } else {
        LockedOverlay(
            requiredTier = requiredTier,
            feature = feature,
            onUpgradeClick = onUpgradeClick,
            modifier = modifier
        )
    }
}

@Composable
private fun LockedOverlay(
    requiredTier: SubscriptionTier,
    feature: String,
    onUpgradeClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Surface(
        modifier = modifier
            .semantics {
                contentDescription = "${requiredTier.displayName} Feature. $feature. Tap to upgrade."
            },
        shape = RoundedCornerShape(14.dp),
        color = MaterialTheme.colorScheme.surfaceVariant,
        border = BorderStroke(1.dp, BrandOrange.copy(alpha = 0.2f))
    ) {
        Column(modifier = Modifier.padding(14.dp)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    imageVector = Icons.Filled.Lock,
                    contentDescription = null,
                    tint = BrandOrange
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    text = "${requiredTier.displayName} Feature",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold
                )
            }

            Spacer(modifier = Modifier.height(4.dp))

            Text(
                text = feature,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )

            Spacer(modifier = Modifier.height(12.dp))

            Button(
                onClick = onUpgradeClick,
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(10.dp),
                colors = ButtonDefaults.buttonColors(
                    containerColor = MaterialTheme.colorScheme.primary
                )
            ) {
                Text(
                    text = "Unlock with ${requiredTier.displayName}",
                    style = MaterialTheme.typography.labelMedium,
                    fontWeight = FontWeight.SemiBold
                )
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun PremiumGateLockedPreview() {
    DesMoinesInsiderTheme {
        PremiumGate(
            requiredTier = SubscriptionTier.INSIDER,
            feature = "Advanced search filters including distance, price range, and rating",
            currentTier = SubscriptionTier.FREE,
            onUpgradeClick = {}
        ) {
            Text("Premium Content Here")
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun PremiumGateUnlockedPreview() {
    DesMoinesInsiderTheme {
        PremiumGate(
            requiredTier = SubscriptionTier.INSIDER,
            feature = "Advanced filters",
            currentTier = SubscriptionTier.INSIDER,
            onUpgradeClick = {}
        ) {
            Text("Premium Content Visible!")
        }
    }
}
